lib/Monads/trace/Trace_RG.thy

```diff
@@ -230,6 +314,135 @@ lemma bind_twp[wp_split]:
   apply (clarsimp simp: guar_cond_append)
   done
 
+lemma bindE_twp[wp_split]:
+  "\<lbrakk>\<And>r. \<lbrace>Q' r\<rbrace>,\<lbrace>R\<rbrace> g r \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>; \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q'\<rbrace>,\<lbrace>E\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f >>=E (\<lambda>r. g r) \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  apply (simp add: bindE_def validIE_def)
+  apply (erule bind_twp[rotated])
+  apply (clarsimp simp: lift_def throwError_def split: sum.splits)
+  apply (subst validI_valid_no_trace_eq)
+   apply wpsimp+
+  done
+
+\<comment> \<open>FIXME MC: improve naming for bind_twp vs seq_ext vs rg_seq_ext
+          existing proofs mention hoare_seq_ext a lot and seq_ext a few times\<close>
+\<comment> \<open>FIXME MC: names of seq_ext and seq_extE clash with valid equivalents\<close>
+lemmas seq_ext = bind_twp[rotated]
+lemmas seq_extE = bindE_twp[rotated]
+
+lemmas rg_seq_ext = bind_twp
+lemmas rg_seq_extE = bindE_twp
+
+lemma rg_TrueI[simp]:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f -,\<lbrace>\<lambda>_ _. \<top>\<rbrace> = prefix_closed f"
+  by (simp add: validI_def)
+
+lemma rgE_TrueI[simp]:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f -,\<lbrace>\<lambda>_ _. \<top>\<rbrace>,\<lbrace>\<lambda>_ _. \<top>\<rbrace> = prefix_closed f"
+  by (simp add: validIE_def)
+
+lemma rg_post_conj[intro]:
+  "\<lbrakk>\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>; \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q'\<rbrace>\<rbrakk> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q and Q'\<rbrace>"
+  by (fastforce simp: validI_def)
+
+lemma rg_pre_disj[intro]:
+  "\<lbrakk>\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>; \<lbrace>P'\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>\<rbrakk> \<Longrightarrow> \<lbrace>P or P'\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  by (fastforce simp: validI_def pred_disj_def)
+
+lemma rg_conj:
+  "\<lbrakk>\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>; \<lbrace>P'\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q'\<rbrace>\<rbrakk> \<Longrightarrow> \<lbrace>P and P'\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q and Q'\<rbrace>"
+  unfolding validI_def by auto
+
+lemmas wp_post_tautI = rg_TrueI[where P="\<top>\<top>", THEN iffD2]
+lemmas wp_post_tautEI = rgE_TrueI[where P="\<top>\<top>", THEN iffD2]
+
+lemma rg_pre_cont[simp]:
+  "prefix_closed f \<Longrightarrow> \<lbrace>\<bottom>\<bottom>\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
```

Comment:
similar to the other two above -- `R` could end up as free schematic (assuming we're in a `wp` setting `G` and `Q` should be concrete).